### ISSUE TITLE
[13] Component: Typeahead Box

### ISSUE THREAD

**Issue by tosbourn** [WeAreSnook/connected-camden#44]
We need a typeahead component to allow a user to enter information into a box, and for clickable suggestions to appear below the box as a list.

Inserting correct data into the component is out of scope for this ticket, (i.e, we don't need to make API calls)

## Acceptance Criteria

- [x] Can call the component and have an input box appear on screen
- [x] Entering text into the input box will generate a list under the box with clickable elements
- [x] It is expected to have a hidden field for latitude
- [x] It is expected to have a hidden field for longitude

**Comment by francosta** [comment]
I've changed the implementation of the autocomplete to use alphagov's [accessible-autocomplete](https://github.com/alphagov/accessible-autocomplete) on @mathewtrivett's suggestion. The implementation was working when tested in a view (`views/trip_builder/starting_point.html.erb`).

After adding the `TypeaheadBox` component using `ViewComponent`, I am having trouble loading the javascript on tests.

@mathewtrivett has implemented all the settings on adding javascript sidecar assets to `ViewComponent` as explained [here](https://viewcomponent.org/#sidecar-assets-experimental) but the javascript doesn't seem to be loading. Testing it n the view, the javascript is not being loaded and we're getting an error `window.autocomplete is not a function`.

The ongoing work has been pushed to the `44-typeahead-box` branch.

**Comment by tosbourn** [comment]
Thanks for the detailed explanation @francosta, I'm going to take a look at this later today and once I've got my head around it we can catch up :)

**Comment by tosbourn** [comment]
@francosta I've reviewed the code and think I can see the issue.

Now that we have a JS file within the component, we need to be handling the call from here. It is slightly different to thinking about things in terms of views.

To get _something_ working I added the following to `typeahead_box_component.js`

```
document.addEventListener("turbolinks:load", function() {
  document.querySelectorAll("#autocomplete-container").forEach(node => {
    name = 'starting_point'; // needs to grab the name, probably from the node
    autoComplete(name);
  })
})
```

This hooks into turbolinks (that we are using for general speed improvements) to say

"When everything has loaded, look for any contains that have and ID of `autocomplete-container` and for each of them run the auto complete code.

For my testing I hard-coded the `name`, but we can pass that as a data attribute, so in our component's HTML we could have something like:

```
<div id='autocomplete-container' data-autocomplete-id="<%= @name %>">
```

Then our JS could grab that from `node`

To make this work locally then I used `<%= render(Forms::TypeaheadBoxComponent.new(name: 'starting_point')) %>` in the `starting_point` view file.

Happy to chat about all of this tomorrow, but wanted to get some code written up before we did :)

**Comment by HarryPotia** [comment]
Blocked due to some technical issues, however @francosta is confident it can be fixed in the next day.

**Comment by francosta** [comment]
@tosbourn, thank you.  I have been able to add the event listener on the `typeahead_box_component.js` file while removing the `script` from the `erb` file and have the JS run fine in the browser - that works fine.

I'm still having issues with testing, though. It seems that when I run `rspec`, the JS doesn't run. I guess what we would need is something like (execute_script)[https://www.rubydoc.info/github/jnicklas/capybara/Capybara%2FSession:execute_script] to run in the test, after we render the component with `render_inline`.

There's also the question of what we would want the `spec` file to test. In my opinion, we would only want to test that the javascript runs when the component is loaded and not really the implementation of the javascript itself, right? If we're testing the actual typeahead, then we'd be testing the library itself, which I don't think is something we'd want to do (?).
We would only want to test: (1) does the script run when the component loads and (2) can we see an `input` field in the screen as a result.

If so, how can we test both of these using Capybara? Happy to chat about it if it's easier. And thank you again 🙌

**Comment by tosbourn** [comment]
By default JavaScript doesn't run in tests, this is because getting the DOM set up and allowing JS execution is a little slower than not doing any of that stuff.

As per [capybara's documentation](https://github.com/teamcapybara/capybara#selenium), Selenium is one of the recommended drivers for doing this. It supports JavaScript and many other things.

I haven't personally set this up in newer versions of Rails - I will have a look today and get back, so as not to be a blocker.

To your point, `execute_script` is excellent when you have a one off thing to run, but doesn't mimic actual browser behaviour.

**Comment by tosbourn** [comment]
@francosta I've added 2 commits to your branch, which hopefully unblock.

https://github.com/WeAreSnook/connected-camden/commit/d5d36b1561a2181e166f5e65cf03f2810119f9d4 - just adds something that is handy for capybara test debugging, now when a feature test fails you can open the last page capybara "sees" as an HTML screenshot.

https://github.com/WeAreSnook/connected-camden/commit/4a67d959a780542c969e05c1c630b5646fc9bf90 - adds in a JS test, I see we've already set up selenium (thanks @mathewtrivett!) so all we needed to do was tell rspec to use it, using `js: true` in the `describe` block.

If you haven't already, run the steps outlined in the README to get the webdriver installed locally.
